(Alt.) Refactor BoundColumn attributes to allow override of class names. #349

graup opened the pull request:
Related issue: #349
Original PR: #367

So now I added `Table.get_column_class_names` which gets the original set of class names (parsed from column attributes) as well as the bound column to replicate the old behavior but allow to override it.

This might still need a bit more documentation, I think.

jieter commented on django_tables2/columns/base.py:
```diff
@@ -282,33 +282,42 @@ def attrs(self):
+
+        return attrs
+
+    def get_td_class_name(self, td_attrs):
```
shouldn't this be `get_td_class_names` or just `get_td_class`?

jieter commented on django_tables2/tables.py:
```diff
@@ -565,6 +565,24 @@ def template(self):
+        Returns a set of HTML class names for cells of a bound column
+        in this tables.
+        By default it uses the class names defined in the table's attributes,
+        and additionally the bound column's name.
```
I think it might help to mention explicityly that these class attribute will be added to both the `<th>` and the `<td> tag.

jieter commented on django_tables2/columns/base.py:
```diff
@@ -282,33 +282,42 @@ def attrs(self):
+        Returns the HTML class attribute for a header cell in this column
+        """
+        classes = set((c for c in th_attrs.get('class', '').split(' ') if c))
+        th_classes = self.table.get_column_class_names(classes, self)
```
Since these are now in a separate function, I think we can remove the `th_`-prefix or use that everywhere.

jieter commented:
I think I like this one better, API is simpler.

Not totally sure about:
- the name of `get_column_class_names`, 
- in the case of the `<th>`, a note about the adding addition of `ordering/descending/ascending` to the class might be in place.
- while writing the previous point, I noticed we have a different but similar way in which we allow customisation of the classes for ordering. Why not use something like that to tackle this problem? (Not very well thought through yet).
- As for documenting it, I think the `get_colum_class_names` docstr should mention it is intended to be overridden to allow customisation.
- Some others which I commented on inline.

graup commented:
I fixed the things you mentioned and added more documentation.

I agree with your third point, it is still a bit weird to have a completely different API for this. I was also thinking about how we could get this into the attributes, but the only option I saw would be to allow passing a function (or lambda) into that, which is also a bit messy.

graup commented:
At least with this version the (eventual) removal of the name class and migration path is very clear. We just have to remove the line `classes_set.add(bound_column.name)` and tell people to override the function to add it back.

jieter commented:
Decided to merge, thanks @graup!

jieter commented:
I just did a pre-release of 2.0.0 with version 2.0.0a0. Please help by testing and report any issues.
